mlflow/tracking/client.py

```diff
@@ -1418,8 +1419,200 @@ def _is_plotly_figure(fig):
+
+        .. code-block:: python
+            :caption: Time-stepped image logging pillow example
+
+            import mlflow
+            from PIL import Image
+
+            image = Image.new("RGB", (100, 100))
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, "dogs", image, step=3)
+
+        .. code-block:: python
+            :caption: Legacy artifact file image logging numpy example
+
+            import mlflow
+            import numpy as np
+
+            image = np.random.randint(0, 256, size=(100, 100, 3), dtype=np.uint8)
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, image, "image.png")
+
+        .. code-block:: python
+            :caption: Legacy artifact file image logging pillow example
+
+            import mlflow
+            from PIL import Image
+
+            image = Image.new("RGB", (100, 100))
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, image, "image.png")
+
+        Raises:
+            TypeError: If unexpected parameters are provided. Please refer to the function
+            documentation for required parameters.
+        """
+        import numpy as np
+        from PIL.Image import Image
+
+        image_over_steps_params = [
+            ("key", str),
+            ("image", (np.ndarray, Image)),
+            ("step", int),
+            ("timestamp", int),
+        ]
+        image_over_steps_required = ["key", "image"]
+
+        image_as_artifact_params = [
+            ("image", (np.ndarray, Image)),
+            ("artifact_file", str),
+        ]
+        image_as_artifact_required = ["image", "artifact_file"]
+
+        def _check_types(param_types, *args, **kwargs):
+            # Check args have correct types
+            for i, arg in enumerate(args):
+                if not isinstance(arg, param_types[i][1]):
+                    return False
+            # Check kwargs are correctly named, have correct types, and not duplicated
+            remaining_kwargs = dict(param_types[len(args) :])
+            for key, value in kwargs.items():
+                if key not in remaining_kwargs:
+                    return False
+                if not isinstance(value, remaining_kwargs[key]):
+                    return False
+
+                del remaining_kwargs[key]
+            return True
+
+        def _check_required(param_required, *args, **kwargs):
+            # Checks if all required parameters are present
+            return any(
+                len(args) >= i and all(key in kwargs for key in param_required[i:])
+                for i in range(len(param_required) + 1)
+            )
+
+        if _check_types(image_over_steps_params, *args, **kwargs) and _check_required(
+            image_over_steps_required, *args, **kwargs
+        ):
+            # log an image over steps
+            def _log_image(
```

Comment:
we are nesting many levels now, shall we move this function out and name it as `_log_image_by_step`?